src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -43,29 +44,22 @@ use futures::future::try_join_all;
 
 pub async fn generate_permutation_opt<F>(
     ctx: SemiHonestContext<'_, F>,
-    sort_keys: &[Vec<Replicated<F>>],
-    num_bits: u32,
-    num_multi_bits: u32,
+    sort_keys: &[Vec<Vec<Replicated<F>>>],
 ) -> Result<Vec<Replicated<F>>, Error>
 where
     F: Field,
 {
+    assert_ne!(sort_keys.len(), 0);
     let ctx_0 = ctx.narrow(&Sort(0));
-    assert_eq!(sort_keys.len(), num_bits as usize);
-
-    let last_bit_num = std::cmp::min(num_multi_bits, num_bits);
 
-    let lsb_permutation = multi_bit_permutation(
-        ctx_0.narrow(&BitPermutationStep),
-        &sort_keys[0..last_bit_num.try_into().unwrap()],
-    )
-    .await?;
+    let lsb_permutation =
+        multi_bit_permutation(ctx_0.narrow(&BitPermutationStep), &sort_keys[0]).await?;
 
     let input_len = u32::try_from(sort_keys[0].len()).unwrap(); // safe, we don't sort more that 1B rows
 
     let mut composed_less_significant_bits_permutation = lsb_permutation;
-    for bit_num in (num_multi_bits..num_bits).step_by(num_multi_bits.try_into().unwrap()) {
-        let ctx_bit = ctx.narrow(&Sort(bit_num));
+    for (bit_num, one_slice) in sort_keys.iter().enumerate().skip(1) {
```

Comment:
`bit_num` isn't a good variable name, because you're enumerating the "slices", so it's like a "chunk_number" now.

Also, `one_slice` is an odd name. Why one? Perhaps `next_few_bits`?

Reply:
I was thinking this is one slice which will be sent to bit permutation at a time. Changing it